pycbc/types/array_cupy.py

```diff
@@ -0,0 +1,140 @@
+# Copyright (C) 2012  Alex Nitz
+# This program is free software; you can redistribute it and/or modify it
+# under the terms of the GNU General Public License as published by the
+# Free Software Foundation; either version 3 of the License, or (at your
+# option) any later version.
+#
+# This program is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
+# Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, write to the Free Software Foundation, Inc.,
+# 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+
+
+#
+# =============================================================================
+#
+#                                   Preamble
+#
+# =============================================================================
+#
+"""Numpy based CPU backend for PyCBC Array
+"""
+import cupy as _xp
+from pycbc.types.array import common_kind, complex128, float64
+from scipy.linalg import blas
+from pycbc.types import real_same_precision_as
+
+def zeros(length, dtype=_xp.float64):
+    return _xp.zeros(length, dtype=dtype)
+
+def empty(length, dtype=_xp.float64):
+    return _xp.empty(length, dtype=dtype)
+
+def ptr(self):
+    return self.data.data.mem.ptr
+
+def dot(self, other):
+    return _xp.dot(self._data,other)
+
+def min(self):
+    return self.data.min()
+
+def abs_max_loc(self):
+    if self.kind == 'real':
+        tmp = abs(self.data)
+        ind = _xp.argmax(tmp)
+        return tmp[ind], ind
+    else:
+        tmp = self.data.real ** 2.0
+        tmp += self.data.imag ** 2.0
+        ind = _xp.argmax(tmp)
+        return tmp[ind] ** 0.5, ind
+
+def cumsum(self):
+    return self.data.cumsum()
+
+def max(self):
+    return self.data.max()
+
+def max_loc(self):
+    ind = _xp.argmax(self.data)
+    return self.data[ind], ind
+
+def take(self, indices):
+    return self.data.take(indices)
+
+def weighted_inner(self, other, weight):
+    """ Return the inner product of the array with complex conjugation.
+    """
+    if weight is None:
+        return self.inner(other)
+
+    cdtype = common_kind(self.dtype, other.dtype)
+    if cdtype.kind == 'c':
+        acum_dtype = complex128
+    else:
+        acum_dtype = float64
+
+    return _xp.sum(self.data.conj() * other / weight, dtype=acum_dtype)
+
+def abs_arg_max(self):
+    if self.dtype == _xp.float32 or self.dtype == _xp.float64:
+        return _xp.argmax(abs(self.data))
+    else:
+        return abs_arg_max_complex(self._data)
+
+def inner(self, other):
+    """ Return the inner product of the array with complex conjugation.
+    """
+    cdtype = common_kind(self.dtype, other.dtype)
+    if cdtype.kind == 'c':
+        return _xp.sum(self.data.conj() * other, dtype=complex128)
+    else:
+        return inner_real(self.data, other)
```

Comment:
same here - i dont see where this is defined

Reply:
Same as above.